tests/unit/test_mysqlsh_helpers.py

```diff
@@ -148,6 +149,112 @@ def test_configure_mysql_users_exception(self, _run_mysqlcli_script):
+                    "mysql-0": {
+                        "status": "online",
+                        "address": "mysql-1.mysql-endpoints",
+                    },
+                    "mysql-2": {
+                        "status": "unreachable",
+                        "address": "mysql-2.mysql-endpoints",
+                    },
+                    "mysql-1": {
+                        "status": "(missing)",
+                        "address": "mysql-1.mysql-endpoints",
+                    },
+                },
+            },
+        }
+        _get_cluster_status.return_value = get_cluster_status_return
+
+        _expected_force_quorum_commands = "\n".join(
+            (
+                "shell.connect('clusteradmin:clusteradminpassword@127.0.0.1')",
+                "cluster = dba.get_cluster('test_cluster')",
+                "cluster.force_quorum_using_partition_of('clusteradmin@mysql-1.mysql-endpoints', 'clusteradminpassword')",
+            )
+        )
+
+        _expected_remove_instance_one_commands = "\n".join(
+            (
+                "shell.connect('clusteradmin:clusteradminpassword@127.0.0.1')",
+                "cluster = dba.get_cluster('test_cluster')",
+                'cluster.remove_instance(\'mysql-1.mysql-endpoints\', {"force": "true"})',
+            )
+        )
+        _expected_remove_instance_two_commands = "\n".join(
+            (
+                "shell.connect('clusteradmin:clusteradminpassword@127.0.0.1')",
+                "cluster = dba.get_cluster('test_cluster')",
+                'cluster.remove_instance(\'mysql-2.mysql-endpoints\', {"force": "true"})',
+            )
+        )
+
+        # disable tenacity retry
+        self.mysql.remove_instances_not_online.retry.retry = tenacity.retry_if_not_result(
+            lambda x: True
+        )
+
+        self.mysql.remove_instances_not_online()
+
+        self.assertEqual(_run_mysqlsh_script.call_count, 3)
+
+        self.assertEqual(
+            sorted(_run_mysqlsh_script.mock_calls),
+            sorted(
+                [
+                    call(_expected_force_quorum_commands),
+                    call(_expected_remove_instance_one_commands),
+                    call(_expected_remove_instance_two_commands),
+                ]
+            ),
+        )
+
+    @patch("mysqlsh_helpers.MySQL.get_cluster_status")
+    @patch("mysqlsh_helpers.MySQL._run_mysqlsh_script")
+    def test_remove_instances_not_online_exception(self, _run_mysqlsh_script, _get_cluster_status):
+        """Test an exception while executing remove_instances_not_online."""
+        # disable tenacity retry
+        self.mysql.remove_instances_not_online.retry.retry = tenacity.retry_if_not_result(
+            lambda x: True
+        )
+
+        _get_cluster_status.return_value = None
+        with self.assertRaises(MySQLRemoveInstancesNotOnlineRetryError):
+            self.mysql.remove_instances_not_online()
+
+        get_cluster_status_return = {
```

Comment:
might be worth storing in a file and re-loading

Reply:
moving to a file may be overkill for now, but moved it to a constant at the top of file in a9bd5bb1a336943a9c0d601492f594aaddd073e8 (to be reused)